### ISSUE TITLE
ReduceLROnPlateau fails for negative input

### ISSUE THREAD

**Issue by jenuk** [pytorch/pytorch#47513]
## 🐛 Bug

The formula in `torch.optim.lr_scheduler.ReduceLROnPlateau` ([documentation](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.ReduceLROnPlateau)) that measures improvement is wrong for a negative metric in relative mode. This forumla is in the method `is_better`.

Specifically in relative mode, if there is a worse performance than before, the formula may return that the performance improved.

## To Reproduce

1. Initialize scheduler with options `mode='min'`, `threshold=0.1`, `patience=0` and `threshold_mode='rel'`
1. call `step` with initial negative value `a`. E.g. `scheduler.step(-1)`
1. call `step` with higher value `b` than before. `b`should be in the interval `(0.9*a, a)`. E.g. `scheduler.step(-.95)`
1. Notice no change in the learning rate despite worse performance and no patience.

```
import torch

model = torch.nn.Linear(1, 1)
opt = torch.optim.SGD(model.parameters(), lr=0.1)
scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(opt, mode="min", factor=0.1, threshold=0.1, patience=0, threshold_mode='rel', verbose=True)

# status
print(opt.param_groups[0]["lr"])
# >>> 0.1

# do epoch 1
scheduler.step(-1.0)

# do epoch 2
scheduler.step(-0.95)

# should be 0.01
print(opt.param_groups[0]["lr"])
# >>> 0.1
```

(Analogous for `mode=max`)

## Expected behavior

`ReduceLROnPlateau.is_better(self, -0.95, -1)` should return `False`. Which would result in `opt.param_groups[0]["lr"] == 0.01` in the example above (barring numeric inaccuracies).

## Environment

PyTorch version: 1.7.0+cpu
Is debug build: True
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Home
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 950M
Nvidia driver version: 388.57
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.17.0+mkl
[pip3] torch==1.7.0+cpu
[pip3] torchaudio==0.7.0
[pip3] torchtext==0.6.0
[pip3] torchvision==0.8.1+cpu
[conda] Could not collect

## Additional context

For positive numbers relative increase can be described by `(current - previous)/previous` (multiply by -1 for decrease in case of minimization), from which the formular in `is_better` is derived, however this does not work if `previous` is negative. The relative change should instead be: `(current - previous)/|previous|`.

In the [source code (line 662)](https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py) this is represented by (Case `mode='min' and threshold_mode == 'rel'`)
```
rel_epsilon = 1. - self.threshold
return a < best * rel_epsilon
```
which should rather be
```
return best - a > self.threshold * abs(best)
```

And analogous for the case `mode='max' and threshold_mode == 'rel'`
```
return a - best > self.threshold * abs(best)
```

This behavior was already noticed in #38622 which believes this to be an error in the documentation rather than the code.

**Comment by mruberry** [comment]
Thank you for reporting this issue, @jenuk. We would accept a PR correcting the documentation.

**Comment by jenuk** [comment]
I don't believe the documentation is the issue here. But I would gladly provide a PR fixing the formula (thus making the documentation accurate). I want to note that I'm not the only one that believes in measuring the relative change in this way, see for example the [corresponding wikipedia article](https://en.wikipedia.org/wiki/Relative_change_and_difference#Definitions), they point out the same issue as I have and offer the same formula.

If you only want to change the documentation maybe @yks72p from the other issue is willing to help.

**Comment by mruberry** [comment]
Aha, I see. Thank you for correcting me, @jenuk, I read your comment too quickly. 

I'll let @vincentqb comment on this issue since it's more than a documentation change.

**Comment by vincentqb** [comment]
I agree that this example contradicts the documentation, and it makes sense to address this issue. Concerns I have are around: how widely users lived with this behavior where a change would be BC breaking (e.g. loss goes accidentally slightly negative? users want a loss near zero and understood "better" as near zero?), and silent change in behavior by silently replacing the denominator with an absolute value. Options I see:

1. Replace denominator with absolute value. Add a warning when a negative value of `previous` is detected when `threshold_mode="rel"` is used. (Add a test showing that the warning is raised, and that the negative case for min/max is handled properly.)
1. Add a new `threshold_mode="rel_abs"` with that behavior. Update the documentation contrasting the `rel` and `rel_abs`.  (Add a test showing how the negative case for min/max are handled for `rel_abs` and for `rel`.)
1. Update the documentation and not change the code. (Add a test showing how the negative case for min/max is handled.)

My preferred solution is 2. Thoughts?

**Comment by jenuk** [comment]
> users want a loss near zero and understood "better" as near zero?

So I actually have a loss that is negative and should be as close to zero as possible. But both `mode='min'` and `mode='max'` do not work for this. If my previous result is `a` then the accepted interval (all results accepted as better) is `(-infinity, (1-threshold)*a)` for minimization and `((1+threshold)*a, infinity)` and both accept some values below and above `a` itself.

My solution is to pass the negative loss to the scheduler and minimize that. This wouldn't break when the behavior for negative loss is changed.


I can't really imagine a situation in which somebody noticed this behavior and used it intentionally, because no matter were the goal is better and worse loss is accepted. However I don't have much experience with software of this scale, so this shouldn't count for too much. 

In any case solution 2 solves my issue with this class.

**Comment by BrunoBelucci** [comment]
Hello,

I have recently faced the same issue when using a loss that could have negative values. As for today, the documentation is still misleading and I agree with @jenuk that I do not see any case in which somebody used this behavior intentionally. On the contrary, I can see many cases where people have either not identified that the code does not work as expected for negative values or that people found out that the code did not work as expected and had to adapt. If the user wants a loss to be "better" near zero, we should implement this behavior both for the 'rel' and 'abs' threshold_mode, because, at least for me, it does not make sense to have this behavior only for the 'rel' threshold_mode. 

My thoughts:
1 - fix 'rel' threshold_mode, because today the expected behavior is clearly not achieved.
2 - we can warn the user about the change, but I think that it is the code that must be fixed in this case.
3 - if people are interested, implement a mode (and not a threshold_mode) 'zero' (or some other name) that should tell that we are neither minimizing ('min') nor maximizing ('max'), but we want the quantity to be near zero.

I am also available to submit a PR to fix this issue.
